### ISSUE TITLE
[Android] Your app schedules exact alarms without checking whether the SCHEDULE_EXACT_ALARM permission has been granted.

### ISSUE THREAD

**Issue by owenashurst** [thudugala/Plugin.LocalNotification#465]
**Describe the bug**
When publishing release builds to the Google Play Developer Console, I found that it is showing me an error for the SCHEDULE_EXACT_ALARM permission. The exact error is:

```
Your app schedules exact alarms without checking whether the SCHEDULE_EXACT_ALARM permission has been granted. This is causing your app to crash for users on Android 14 because the permission is no longer granted by default.
```

The AndroidManifest has the permissions enabled outlined in the Wiki for this plugin, including the SCHEDULE_EXACT_ALARM permission. When testing via debug, I can see the notification that was scheduled fine (Google Pixel 5 emulator & My own Google Pixel 7 Pro).

**To Reproduce**
1. Install Plugin.LocalNotifications
2. Set a notification to appear by scheduling a DateTime.
3. Publish the release to Google Play Console (Internal testing etc)

**Expected behavior**
I should expect the plugin to have the ability to request missing permissions that it uses internally, instead of just one ability to Request Notification permissions.

**Platform (please complete the following information):**
 - OS: Android
 - Version 14

**Smartphone (please complete the following information):**
 - Device: Google Pixel
 - OS: Android
 - Version 14

**Comment by vallgrenerik** [comment]
Im also facing this! 
Did you @owenashurst  find any solution?
Ping @thudugala

**Comment by owenashurst** [comment]
> Im also facing this! 
> Did you @owenashurst  find any solution?
> Ping @thudugala 

Nope! I read a comment on another issue that a potential fix is in master. There needs to be a new release ideally. I tried cloning the code and including it in the project but I had all sorts of issues.

**Comment by thudugala** [comment]
@owenashurst new releases are planned only to be release using .Net 8 Maui

**Comment by thudugala** [comment]
@owenashurst can release new version, if you do not any issues with latest pre release version

**Comment by owenashurst** [comment]
> @owenashurst can release new version, if you do not any issues with latest pre release version

Just installed the pre-release, and it seems to still work fine (for android at least). I think you're good. I'm ready to deploy a new version of my app, so I'm happy to update once you release so I can publish.

Quick question, are you saying I can remove the following permissions after your release?

``` 
<uses-permission android:name="android.permission.USE_EXACT_ALARM" />
<uses-permission android:name="android.permission.SCHEDULE_EXACT_ALARM" />
```

**Comment by thudugala** [comment]
@owenashurst yes.

But please remember notification will not pop up at the exact time.

Try with the pre release

**Comment by owenashurst** [comment]
> @owenashurst yes.
> 
> But please remember notification will not pop up at the exact time.
> 
> Try with the pre release

@thudugala My app doesn't need to be exact. As long as it shows the notification +-  a few seconds and not minutes.

I've just tried with the above permissions removed and it was still relatively exact. Although this was inside an Android Emulator.

Does that mean it's good?

**Comment by thudugala** [comment]
@owenashurst yes.

**Comment by owenashurst** [comment]
@thudugala Brilliant. Looking forward to the release. Thank you.

**Comment by thudugala** [comment]
Will release tonight at about 10 pm New Zealand time

**Comment by thudugala** [comment]
@owenashurst Please try version [11.1.0](https://www.nuget.org/packages/Plugin.LocalNotification/11.1.0)

**Comment by vallgrenerik** [comment]
Thanks for the update @thudugala 

In the Wiki you have:

```
[assembly: UsesPermission(Manifest.Permission.WakeLock)]

//Required so that the plugin can reschedule notifications upon a reboot
[assembly: UsesPermission(Manifest.Permission.ReceiveBootCompleted)]
[assembly: UsesPermission(Manifest.Permission.Vibrate)]
[assembly: UsesPermission("android.permission.POST_NOTIFICATIONS")]

optional
[assembly: UsesPermission("android.permission.USE_EXACT_ALARM")]
[assembly: UsesPermission("android.permission.SCHEDULE_EXACT_ALARM")]
```

Could you elaborate, when should we use the optional permissions and when should we avoid them? :)

**Comment by thudugala** [comment]
@vallgrenerik if you want notifications to fire at the exact time you need the optional permission.

**Comment by andrea-torre** [comment]
Hi, I'm having this problem too, in Google Play Console I can see the same warning for Android 14 that the original poster described. My App is developed with .net 7.0. It's not clear to me if I have to update to version 11.1.0 and .net 8.0 in order to solve the problem or if it is possible to deal with the permissions also without porting to .net 8.0. Thank you

**Comment by alexsmi-noveo** [comment]
For .NET8.0 Google Play fires me warning: "Your app schedules exact alarms without checking whether the SCHEDULE_EXACT_ALARM permission has been granted. This is causing your app to crash for users on Android 14 because the permission is no longer granted by default". 

I was managed to request SCHEDULE_EXACT_ALARM permission (Android, API 31+) using android native classes (MainActivity.cs):

```
AlarmManager alarmManager = AlarmManager.FromContext(Platform.AppContext);
var isPermissionSet = alarmManager.CanScheduleExactAlarms();

if (!isPermissionSet)
{
    Android.Net.Uri uri = Android.Net.Uri.Parse("package:" + Android.App.Application.Context.ApplicationInfo.PackageName);
    //https://developer.android.com/reference/android/provider/Settings#ACTION_REQUEST_SCHEDULE_EXACT_ALARM  
    Intent intent = new Intent(Android.Provider.Settings.ActionRequestScheduleExactAlarm, uri);
    StartActivity(intent);
}
```

**Comment by Kebechet** [comment]
Except the permissions in `AndroidManifest` consider adding `AndroidNotificationPermission`

```
private static NotificationPermission _notificationPermissions = new()
{
	Android = new AndroidNotificationPermission()
	{
		RequestPermissionToScheduleExactAlarm = true,
	}
};
```

and then call:
```
var areNotificationsEnabled = await notificationService.AreNotificationsEnabled(_notificationPermissions);
await notificationService.RequestNotificationPermission(_notificationPermissions);
```

**Comment by erossini** [comment]
I'm working on a `NET8` `MAUI` app, I can't publish the app on Google Play because I get the same error that @owenashurst had. If I remove the `SCHEDULE_EXACT_ALARM`, the user can't accept the notifications (see this [post](https://github.com/thudugala/Plugin.LocalNotification/issues/503)).

Do you have any suggestions/ideas how can I fix it?

**Comment by Kebechet** [comment]
I had to remove `USE_EXACT_ALARM` because from android 14 this permission in not granted by default and Google allows it only for calendar and alarm apps.

So I kept `SCHEDULE_EXACT_ALARM` and asked for permissions with: https://github.com/thudugala/Plugin.LocalNotification/issues/465#issuecomment-2106834738

**Comment by erossini** [comment]
> I had to remove `USE_EXACT_ALARM` because from android 14 this permission in not granted by default and Google allows it only for calendar and alarm apps.
> 
> So I kept `SCHEDULE_EXACT_ALARM` and asked for permissions with: [#465 (comment)](https://github.com/thudugala/Plugin.LocalNotification/issues/465#issuecomment-2106834738)
> 
> After that Google allowed our release.
> 
> But I still see the warning that for some of our users it can crash the app. I have no idea why because before every notification call we have permission check

Just to clarify for me. You kept in the _AndroidManifest.xml_ the line 

    <uses-permission android:name="android.permission.SCHEDULE_EXACT_ALARM" />

and then add the code in the code above. How did you define `notificationService`?

**Comment by erossini** [comment]
@Kebechet Thank you so much! Very useful!

**Comment by erossini** [comment]
When the app starts, the authorization page is displayed but I can't enable the alarms.

![image](https://github.com/thudugala/Plugin.LocalNotification/assets/9497415/f81cea15-8e28-4e02-945d-7ae4feec0cc9)

**Comment by Kebechet** [comment]
In my app I can enable those alarms. I have all these permissions enabled:
```
	<!--Required so that the plugin can reschedule notifications upon a reboot-->
	<uses-permission android:name="android.permission.WAKE_LOCK" />
	<uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED" />
	<uses-permission android:name="android.permission.VIBRATE" />
	<uses-permission android:name="android.permission.POST_NOTIFICATIONS" />

	<!--Unlike USE_EXACT_ALARM , the SCHEDULE_EXACT_ALARM permission must be granted by the user-->
	<uses-permission android:name="android.permission.SCHEDULE_EXACT_ALARM" />
```
otherwise I dunno what could be wrong with your implementation

**Comment by Kebechet** [comment]
After upgrading to the latest version of the library the error disappeared
